R/get_acoustic_detections.R

```diff
@@ -70,20 +71,140 @@
 get_acoustic_detections <- function(connection,
                                     start_date = NULL,
                                     end_date = NULL,
+                                    detection_id = NULL,
                                     acoustic_tag_id = NULL,
                                     animal_project_code = NULL,
                                     scientific_name = NULL,
                                     acoustic_project_code = NULL,
                                     receiver_id = NULL,
                                     station_name = NULL,
                                     limit = FALSE,
-                                    api = TRUE){
+                                    api = TRUE) {
   # Check arguments
   # The connection argument has been depreciated
   if (lifecycle::is_present(connection)) {
     deprecate_warn_connection()
   }
-  # Either use the API, or the SQL helper.
-  out <- conduct_parent_to_helpers(api)
-  return(out)
+  assertthat::assert_that(assertthat::is.flag(api))
+  assertthat::assert_that(assertthat::is.flag(limit))
+
+  # Some arguments don't need to be send to etnservice
+  arguments_to_pass <-
+    return_parent_arguments(depth = 1)[
+      !names(return_parent_arguments(depth = 1)) %in% c(
+        "api",
+        "connection",
+        "limit" # handled client side
+      )
+    ]
+
+  # Either use the API, or the SQL helper
+  if (api) {
+    n_records_returned <- forward_to_api("get_acoustic_detections_page",
+      payload = append(
+        arguments_to_pass,
+        list(count = TRUE)
+      ),
+      json = TRUE
+    ) %>%
+      dplyr::pull("count")
+  } else {
+    n_records_returned <- do.call(etnservice::get_acoustic_detections_page,
+      args = append(
+        arguments_to_pass,
+        list(count = TRUE)
+      )
+    ) %>%
+      dplyr::pull("count")
+  }
+
+  # Initialise progress bar with total records expected
+  cli::cli_progress_bar(total = n_records_returned)
+
+  # Control number of objects to fetch per page, 100k default, up to 1M for
+  # big queries
+  page_size <- dplyr::case_when(
+    limit ~ 100,
+    n_records_returned > 5e6 ~ 1e6,
+    .default = 100000
+  )
+
+  # Init object to store pages
+  combined_results <- list()
```

Comment:
It's much more efficient to create a list of known length and then subscript to list elements instead of dynamically growing the list. 

Then again, the length of the list is limited so the actual performance hit on this is probably relatively minor.